examples/counter-with-callback/apps/src/main.rs

```diff
@@ -0,0 +1,304 @@
+// Copyright (c) 2025 RISC Zero, Inc.
+//
+// All rights reserved.
+
+use std::time::{Duration, SystemTime};
+
+use crate::counter::{ICounter, ICounter::ICounterInstance};
+use alloy::{
+    primitives::{aliases::U96, utils::parse_ether, Address},
+    signers::local::PrivateKeySigner,
+    sol_types::SolCall,
+};
+use anyhow::{bail, Context, Result};
+use boundless_market::{
+    client::ClientBuilder,
+    contracts::{Callback, Input, Offer, Predicate, ProofRequest, Requirements},
+    storage::{BuiltinStorageProvider, StorageProvider, StorageProviderConfig},
+};
+use clap::Parser;
+use guest_util::{ECHO_ELF, ECHO_ID};
+use risc0_zkvm::{default_executor, sha::Digestible, ExecutorEnv};
+use url::Url;
+
+/// Timeout for the transaction to be confirmed.
+pub const TX_TIMEOUT: Duration = Duration::from_secs(30);
+
+mod counter {
+    alloy::sol!(
+        #![sol(rpc, all_derives)]
+        "../contracts/src/ICounter.sol"
+    );
+}
+
+/// Arguments of the publisher CLI.
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct Args {
+    /// URL of the Ethereum RPC endpoint.
+    #[clap(short, long, env)]
+    rpc_url: Url,
+    /// URL of the offchain order stream endpoint.
+    #[clap(short, long, env)]
+    order_stream_url: Option<Url>,
+    /// Storage provider to use
+    #[clap(flatten)]
+    storage_config: StorageProviderConfig,
+    /// Private key used to interact with the Counter contract.
+    #[clap(long, env)]
+    private_key: PrivateKeySigner,
+    /// Address of the Counter contract.
+    #[clap(short, long, env)]
+    counter_address: Address,
+    /// Address of the SetVerifier contract.
+    #[clap(short, long, env)]
+    set_verifier_address: Address,
+    /// Address of the BoundlessMarket contract.
+    #[clap(short, long, env)]
+    boundless_market_address: Address,
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    tracing_subscriber::fmt()
+        .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+        .init();
+
+    match dotenvy::dotenv() {
+        Ok(path) => tracing::debug!("Loaded environment variables from {:?}", path),
+        Err(e) if e.not_found() => tracing::debug!("No .env file found"),
+        Err(e) => bail!("failed to load .env file: {}", e),
+    }
+
+    let args = Args::parse();
+
+    // NOTE: Using a separate `run` function to facilitate testing below.
+    run(
+        args.private_key,
+        args.rpc_url,
+        args.order_stream_url,
+        BuiltinStorageProvider::from_config(&args.storage_config).await?,
+        args.boundless_market_address,
+        args.set_verifier_address,
+        args.counter_address,
+    )
+    .await?;
+
+    Ok(())
+}
+
+/// Main logic which creates the Boundless client, executes the proofs and submits the tx.
+async fn run<P: StorageProvider>(
```

Comment:
Since a lot of the comments in this function are repeated in the original `counter` example, I think we could remove them and try and make the example as focused on just callbacks as possible (this is what i did in the smart contract requestor example)